update_handlers.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+)
+
+const (
+	mbrPartitionName  = "mbr.img"
+	bootPartitionName = "boot.img"
+	rootPartitionName = "root.img"
+)
+
+// httpHandler handles a http update link.
+func httpHandler(response *updateResponse) (*zip.ReadCloser, io.ReadCloser, io.ReadCloser, io.ReadCloser, error) {
+	// The link may be a relative url if the server's backend registry is its local disk.
+	// Ensure we have an absolute url by adding the base (updateEndpoint) url
+	// when necessary.
+	link, err := ensureAbsoluteHTTPLink(updateEndpoint, response.Link)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("error ensuring absolute HTTP link '%s' + '%s': %w",
+			updateEndpoint, response.Link, err)
+	}
+
+	log.Printf("downloading update file from registry '%s' with url: %s", response.RegistryType, link)
+
+	filePath := "/tmp/drive.zip"
+	if err := httpDownloadFile(filePath, link); err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("unable to download update file: %w", err)
+	}
+
+	log.Print("loading disk partitions from update file")
+
+	r, err := zip.OpenReader(filePath)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("error opening downloaded file %s: %w", filePath, err)
+	}
+
+	var mbrReader, bootReader, rootReader io.ReadCloser
+	for _, f := range r.File {
+		switch f.Name {
+		case mbrPartitionName:
+			mbrReader, err = f.Open()
+			if err != nil {
+				return nil, nil, nil, nil, fmt.Errorf("error reading %s within update file: %w", mbrPartitionName, err)
+			}
+		case bootPartitionName:
+			bootReader, err = f.Open()
+			if err != nil {
+				return nil, nil, nil, nil, fmt.Errorf("error reading %s within update file: %w", bootPartitionName, err)
+			}
+		case rootPartitionName:
+			rootReader, err = f.Open()
+			if err != nil {
+				return nil, nil, nil, nil, fmt.Errorf("error reading %s within update file: %w", rootPartitionName, err)
+			}
+		}
+	}
+
+	return r, mbrReader, bootReader, rootReader, nil
+}
+
+// httpDownloadFile downloads a file from a url to a filepath on disk.
+// It's efficient because it will write as it downloads
+// and not load the whole file into memory.
+func httpDownloadFile(filepath string, url string) error {
+	// Get the data
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Create the file
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	// Write the body to file
+	_, err = io.Copy(out, resp.Body)
+	return err
```

Comment:
Check the error that file.Close() returns:

if err := out.Close(); err != nil {